### PR TITLE
Fixed Verifiy email using regex and modified Login action

### PR DESCRIPTION
**Pull request by flyznex**
### Prerequisites
- Use regex to verify valid email or username
- Change first parameter of PasswordSignInAsync() method in Login action to TUser type
### Description
- If we use username in method PasswordSignInAsync(), This will call FindUserByName() to find user in database. Of course, method FindByEmailOrUsernameAsync() in Userservice already to return user.

### CODE REVIEW

**Comment by FanrayMedia** [comment]
Hi @flyznex you are so fast, before I could even get the [Contributing Guide](https://github.com/FanrayMedia/Fanray/blob/master/CONTRIBUTING.md) ready :)

For the email check, you probably got the code from the [doc here](https://docs.microsoft.com/en-us/dotnet/standard/base-types/how-to-verify-that-strings-are-in-valid-email-format), I originally thought it was too long (#139) so I didn't go for it :)  But I agree this should be used instead, I also liked your use of extension method.  

[On the bottom of that same doc](https://docs.microsoft.com/en-us/dotnet/standard/base-types/how-to-verify-that-strings-are-in-valid-email-format#compiling-the-code) they provided some ready to use testing code with both inputs and outputs (img below). Would you mind add some Unit Tests with those?  Just add a class `RegexUtilitiesTest.cs` in `Fan.UnitTests/Helpers` folder and create one [Theory] method for all of it.

email test cases

And indeed the [PasswordSignInAsync with userName](https://github.com/aspnet/Identity/blob/master/src/Identity/SignInManager.cs#L299) does make an extra call to look up user, since I already got the user no need for that.  Good catch!  Thank you!

**Comment by flyznex** [comment]
I have created RegexUtilitiesTest.cs with 2 method to verify Extension method working.

**Comment by FanrayMedia** [comment]
On a closer look the `RegexUtilities` cannot be static because it depends on a private variable `_invalid`, I updated the test class and you can see why easier.  When you run this theory it executes all the test cases in parallel, if _invalid were static it will be shared and thus get changed by any test thus fail randomly.  Your original test methods were using a looping, so you won't see this effect.

**Comment by flyznex** [comment]
Thanks for your suggest. I have changed `RegexUtilities` by remove depends `_invalid`.

File: src/Fan/Helpers/RegexUtilities.cs [hunk 1]
```diff
@@ -18,15 +22,14 @@ public static bool IsValidEmail(this string strIn)
...
                 strIn = Regex.Replace(strIn, @"(@)(.+)$", DomainMapper,
                                       RegexOptions.None, TimeSpan.FromMilliseconds(200));
+                if(string.IsNullOrEmpty(strIn))
+                    return false;
             }
-            catch (RegexMatchTimeoutException)
+            catch (Exception ex)
             {
-                return false;
+                if(ex is RegexMatchTimeoutException || ex is ArgumentNullException)
```
Comment (FanrayMedia) [inline src/Fan/Helpers/RegexUtilities.cs]: If you are checking for null on line 25, why would you check for `ArgumentNullException` on this line?

Comment (flyznex) [reply src/Fan/Helpers/RegexUtilities.cs]: Oh, I forgot to remove it. My mistake!

**Comment by FanrayMedia** [comment]
Now you got it to work with static, I'm thinking since I already have a `Util.cs` and this functionality is just for checking for a valid email.  Would you mind if I move this method along with the private static `DomainMapper` into `Util.cs`?  

From that point on calling it would just look like `Util.IsValidEmail("")`, instead of `RegexUtilities.IsValidEmail("")`, it feels tidier and with fewer classes floating around.

**Comment by flyznex** [comment]
Yes you can move it to `Util.cs`. If your application have other string process using regex, I think we can keep it separate. You also can change `IsValidEmail(this string strIn)` to make it become extension method. Regards!

**Comment by FanrayMedia** [comment]
Yes we can make it an extension method anytime. And for right now it's probably a bit cleaner keeping fewer classes till till we have more code.  I just merged it to `dev` branch and will push it live later.

Thank you so much for submitting this PR, you are awesome 👍

**Pull request merged**